### ISSUE TITLE
eval-like mechanism to support e.g. ssh-agent

### ISSUE THREAD

**Issue by bbarker** [nushell/nushell#2812]
**Is your feature request related to a problem? Please describe.**

POSIX `eval` allows updating the environment. Typically, `ssh-agent` is started within a shell so that it functions correctly as `eval $(ssh-agent)`, but `eval` does not currently exist in nu.

**Describe the solution you'd like**
I don't have a strong opinion here; the POSIX behavior doesn't seem to be terrible, given the interface of `ssh-agent`.

**Describe alternatives you've considered**
One workaround that works, if running `nu` from a POSIX-compliant shell, is to just run `eval $(ssh-agent)` before starting `nu` (or running `nu eval` in case you really don't want the outer shell to remain running), but this doesn't work if you don't have other shells at hand.

**Additional context**

> `ssh-agent` would need a translation step since it outputs either bash or c-shell. The c-shell one is pretty close to what nushell uses you would need to massage the c-shell version a bit. Replacing set-env where it uses setenv and adding a = before the value. At that point it should run okay.

If this isn't a priority, I'd be happy to try to implement the feature once I get some Rust under my belt (well, that doesn't sound hygienic, but hopefully you know what I mean).

**Comment by Tarnadas** [comment]
`eval` mechanism is also required to use [fnm](https://github.com/Schniz/fnm)

**Comment by DanCardin** [comment]
This seems like it would be required for a lot of different cli tools to be able to hook into/support nu, and therefore probably be a significant blocker for a bunch of people's practical adoption of it.

It seems like adjusting the `source` command to not require a file argument and to be able to run, for example, `echo 'echo foo' | source`. Which seems like it would fit well into nu's general style, and not require a whole separate `eval` command?

**Comment by ajeetdsouza** [comment]
Adding to the list here - a way to `echo 'foo' | source` would allow [zoxide](https://github.com/ajeetdsouza/zoxide) to seamlessly integrate with Nushell.

**Comment by bbarker** [comment]
Still relevant

**Comment by ajeetdsouza** [comment]
Relevant.

**Comment by Nytelife26** [comment]
Here's my quick hack for anyone in the meantime:
`ssh-agent | sd ';.+' '' | lines | first 2 | split column '=' name value | load-env`

Unfortunately, it seems like ssh-agent isn't working with it for me, despite the environment variables being set.

**Comment by jntrnr** [comment]
@Nytelife26 what does `ssh-agent` output for you?

**Comment by Nytelife26** [comment]
> @Nytelife26 what does `ssh-agent` output for you?

```
SSH_AUTH_SOCK=/tmp/PATH; export SSH_AUTH_SOCK;
SSH_AGENT_PID=XXXX; export SSH_AGENT_PID;
echo Agent pid XXXX;
```

I assume that should be the same everywhere.

**Comment by jntrnr** [comment]
Oh I misread your previous comment. Glad you got it working!

**Comment by Nytelife26** [comment]
> Oh I misread your previous comment. Glad you got it working!

No worries.

What I've found is that nushell's internal commands are so powerful that often you can map external commands into a format you can pipe into nushell to make it do what you want, so an eval is rarely necessary. The UNIX philosophy dictates that commands shouldn't do more than they're designed to, so it would make more sense if `ssh-agent` only outputted the new socket and PID in general, which would make mapping it much simpler. Eval may be nice to have if more programs adopt the ability to generate scripts in nushell's format, particularly with command completions, but that's out of scope here.

Thank you for attempting to assist even though I had no problem, truly kind of you.

**Comment by fdncred** [comment]
Here's a couple different solutions from discord.

from jt
```
> let ssh_env_vars = (ssh-agent -c | lines | str trim --right -c ';' | first 2 | split column " " setenv name value | reject setenv | reduce --fold {} {|it, acc| $acc | insert $it.name $it.value } )
> load-env $ssh_env_vars
```


from ejr
```
# ssh-agent setup
let agentvars = (ssh-agent -c | lines | first 2 |parse "{cmd} {name} {value};" | reject cmd)
let-env $agentvars.0.name = $agentvars.0.value
let-env $agentvars.1.name = $agentvars.1.value
ssh-add ~/.ssh/id_ed25519
```

**Comment by Nytelife26** [comment]
> In 0.60.0, @Nytelife26's ssh-agent solution no longer works:

Here's my new adapted version.

```
ssh-agent -c | sd 'setenv ([\w_]+) (.+);' '$1 $1' | lines | first 2 | split column ' ' | reduce -f {} { |it, acc| $acc | upsert $it.column1 $it.column2 } | load-env
```

**Comment by kubouch** [comment]
We don't plan to have an `eval` functionality as it would significantly interfere with our Nushell's design. That's one limitation of Nushell -- you could say that Nushell doesn't support `eval` by design.

 I'll close this issue since there seems to be a working solution for `ssh-agent`. I would encourage to submit this workaround to [our Cookbook](https://www.nushell.sh/cookbook) so it's easier to find for future users.

**Comment by Fishrock123** [comment]
Also running into a use case where I am having trouble adapting a coworker's `zsh` command to nushell which use `eval`... it would be nice to have this...

**Comment by Nytelife26** [comment]
> Also running into a use case where I am having trouble adapting a coworker's `zsh` command to nushell which use `eval`... it would be nice to have this...

It's almost certainly possible to do what you're looking for without using `eval`
If you tell us about the situation, we can probably help you come up with a better way of doing it in nushell

**Comment by Fishrock123** [comment]
Ended up doing something like you did above, piping into `let-env` with parsing... seems to work ok.

**Comment by dhilipsiva** [comment]
For anyone visiting this thread in the future, it is already added in the cookbook here: https://www.nushell.sh/cookbook/misc.html#manage-ssh-passphrases

```
ssh-agent -c | lines | first 2 | parse "setenv {name} {value};" | transpose -i -r -d | load-env
```

**Comment by koalazub** [comment]
> For anyone visiting this thread in the future, it is already added in the cookbook here: https://www.nushell.sh/cookbook/misc.html#manage-ssh-passphrases
> 
> ```
> ssh-agent -c | lines | first 2 | parse "setenv {name} {value};" | transpose -i -r -d | load-env
> ```

Page has been moved to:
https://www.nushell.sh/cookbook/ssh_agent.html#manage-ssh-passphrases

**Comment by intellild** [comment]
My solution for [keychain](https://www.funtoo.org/Funtoo:Keychain) users
```
load-env (keychain --eval --quiet id_ed25519 | lines | split column ";" | get column1 | first 2 | split column "=" | rename name value | reduce -f {} {|it, acc| $acc | upsert $it.name $it.value })
```

**Comment by Notarin** [comment]
Hey, this might be closed, but as @ajeetdsouza pointed out, this is an issue for zoxide as well.
The current solution requires writing the output to a file then sourcing that file.
The notion that "You can run script, but it absolutely has to come from the filesystem and no other source." is wild to me.
The current required solution is to write the output to a file, source the file, and then be completely done with the file. That is wild and just not acceptable, I would understand if it was for files that have long lifetimes and other uses or whatever, but in this context we're supposed to write a file that will be used a millisecond later and then be finished with, resulting in a file that only ever existed to be useful for a single millisecond...

Not only this, but I sense that this **will** be an issue for more later in other contexts. What if the environment whatever script is ran in does not have permission to write to the fs? What if the environment it is ran in doesn't have a relevant fs? What if a constraint of whatever project requires the script be a single file (see below)?
Let me elaborate on why that would be an issue, when a script is ran by nu, all the source lines are evaluated before the first line of script is ran, meaning that the file cannot exist yet because the code to generate it has not been executed.
aamof, I suspect that this is recursive, meaning the **only** way to actually do this is to put the line that generates the file in `env.nu`, and source it in `conf.nu`, which has its own whole nest of issues. This is present in the presented way to bootstrap zoxide, requiring you generate the file in `env.nu`, which is just not what `env.nu` is meant for.

Sourcing, or evaluating, or running, call it whatever, is an in-shell task. Unless something in the code is **explicitly** causing side-effects, then the shell/process keep its little hands to itself, staying in the RAM where it belongs, except it doesn't, it for what I believe is little to no good reason is **required** to read, and possibly write to, the filesystem, which is *incredibly* slower than handling RAM. It is, internally to the process, such a massive gymnastic maneuver to do what could be far far simpler. It is just such a weird escalation of requirements for such basic functionality. Genuinely the idea of making burner files for every single eval bothers me so much I might write a utility program specifically to utilize FIFO to make this less absurd, I just hope I can look in manfiles and find if there's some service or medium that has a pre-existing socket that I can utilize so I wouldn't have to make a new one.

@kubouch, you said "We don't plan to have an eval functionality as it would significantly interfere with our Nushell's design.", but this is arguably just untrue. Nushell has an eval functionality, it just requires the code to be pre-existing on the filesystem. I find your stance short-thought, because:
1. The functionality almost exists, but the requirement of it being in a file stops the possibility, which feels like such a *small* thing preventing a big set of functionality, functionality required to be a fully operable shell.
2. I suspect that the "design" you talk of is nushells pipeline philosophy, being somewhat similar to the functional paradigm (as in direction of data and being self-contained), which I feel is incorrect as writing to fs for a tiny operation is such a huge side-effect.

I really hope you would be willing to re-open this issue and further explore solutions to making this better. I am willing to discuss possible implementations that do not insult the design philosophy of nu, hell, I'm willing to maybe even contribute the functionality myself if I can.

P.S. Please don't misread my tone, I am not ill or mean, I just stress my points. I genuinely hope for constructive conversation.

**Comment by DanCardin** [comment]
More generally, there is a common pattern of tools exposing a subcommand/flag that outputs shell specific initialization behavior. Without this sort of feature, the situation that played out in this thread (a bunch of people individually posing solutions for 1 specific tool, and everyone needing to copy-paste that magic invocation around to their own config) is going to happen for every tool, rather than the tool author being able to ship native support for nu. Or rather, the tools can perhaps ship support for nu, but the UX will be objectively worse than it is for every other shell.

And fwiw, I've long agreed with Notarin's take. I fail to understand how a command that accepts nu source from a byte stream is meaningfully different from one that accepts it from a file on disk (through a stream of bytes).

**Comment by davidkna** [comment]
I believe that `source` is not `eval`-like in nushell, and instead processed in as static manner before running anything else as some sort of static include. External commands that return initialization commands are not available in that context. If not for that, something like `psub` could be used.

**Comment by kubouch** [comment]
Thanks for the comments, we recently updated our doc pages explaining the eval-related topics: https://www.nushell.sh/book/thinking_in_nu.html#think-of-nushell-as-a-compiled-language and https://www.nushell.sh/book/how_nushell_code_gets_run.html. I encourage you to read those to understand our reasoning. In short, Nushell has more common with statically compiled languages like Rust or C than with dynamic languages like Python or shells. Asking for `eval` in Nushell is similar to asking for `eval` in Rust or C. If we added `eval`, Nushell would become a different type of language.

The absence of `eval` means that once a file is parsed, the resulting intermediate representation (IR) is guaranteed to not change at runtime. You can prepare a different file, then source it later, but it needs to be in a new parse-eval cycle. That's why you need to generate the zoxide init `env.nu`, then source it in `config.nu` because it's run as parse+eval `env.nu`, then parse+eval `config.nu`. File writing happens at the eval stage, that's why you need a new parse-eval cycle. You could also do it in two REPL lines -- each REPL invocation is a new parse-eval cycle.

This is annoying, but adding `eval` would require making compromises in areas we think are important. For example, static analysis tools would instantly become less reliable. That doesn't mean just LSP, but also internal Nushell's stuff like syntax highlighting or completions. Error reporting would also suffer, and many parse-time errors would become runtime errors. We would need some sort of GIL to deal with parallelism, etc. All these things rely on the fact that Nushell's IR is immutable at runtime.

Note that we've been adding features to partially mitigate the pains of not having `eval`:
* Parse-time evaluation and the `const` keyword
* Autoloading https://github.com/nushell/nushell/pull/13217. I believe zoxide could utilize this one.
* Hooks can also circumvent some eval-related limitations.
These are not 100% complete and can be improved as necessary.

Yet another workaround that could be considered is “linking”, where you'd provide a static parse-time known signature of a command, but the body would be supplied at runtime, just like linking a C library.

And finally, some problems do not actually _require_ eval, and you just need to look at the problem from a different angle. A good example is `ssh-agent` which can be handled with `load-env` (I could imagine something like [this](https://www.nushell.sh/cookbook/ssh_agent.html#workarounds) in our standard library).

We know that this is a significant limitation and try to work around the issues people are encountering, but at the same time, the cons outweigh the pros, and we have no plans for adding `eval` at the moment.

**Comment by Notarin** [comment]
@kubouch I really appreciate you making the logic behind your decision transparent, it makes the decision phenomenally easier to understand and get behind.
Furthermore, given your explanation, I agree! I am an LSP addict and I hate the idea of symbol analysis being compromised. Also given further thought I now disagree on the requirement of `eval`, I appreciate the pseudo-compiled nature of nushell. My new opinion is that tools should conform to a format where their actions should be outputting a list of their effects (like env vars being set), and its up to the user of the tool to load the dictionary into their shell (e.g. current ssh-eval solution, but would be better if the output of ssh-eval was more data oriented).

However, this still leaves a problem with zoxide, as well as possibly more tools. Zoxide sets aliases and defines functions, is there a way to load a data source and its contents be setting an alias? Like could I run `fancy_tool_generate_aliases` which outputs:
```
aliases {
  foo=bar
  bazz=idk
}
another-property-for-examples-sake {
  whatever=whenever
}
```
And taking that, I do whatever which each sets of properties, but I also load the aliases into the user environment.
You may want to quickly draw the response that that also breaks static analysis, however I think that the compiled idea kinda falls apart a bit when we're talking about a interactive shell and its utility functions designed explicitly for use in the interactive session, not that it's flawed in general, I just think it has pitfalls in a couple areas.
However, I actually don't really think we need to even sacrifice static analysis, here's my idea:
I think that aliases should be a interactive first thing. I think that what should be done, is that the entire nu file (the config in this case) and all its dependants should be treated as one big evaluation, aka compiled, and that in the context of that file, set aliases should not take effect, and should instead be applied and usable *after* the context of the file. Additionally, I think some function like `load-into-aliases` should exist, also taking effect after the file is evaluated.
This would result in an even stronger than before static analysis, as well as resolving pretty much any issue I can think of in the eval absence. Again, the reason I think that this is needed is because nushell is an interactive shell, and needs the ability to manipulate the pleasantries of the shell (e.g. aliases) live.

If this is implemented, hopefully in a way that is cohesive, I plan to make an issue with zoxide to update the `zoxide init nushell` command to output a more data-oriented output.

I really think my idea here is both compiled friendly, and data oriented, I think it is perfect for nushell.

**Comment by kubouch** [comment]
I'm glad I helped you see the hidden strength of Nushell :-).

Aliases define declarations, similar to `def`, i.e., they add stuff to the IR. So it's not possible to define an alias coming from `zoxide init` at runtime. Doing so would also count as “eval”. I used “eval” in the above explanation for modifying parser definitions or the IR at runtime in general, it is not limited only to the `eval` keyword sourcing some code string.

**Comment by Notarin** [comment]
Heck, furthermore I think I might open a PR with openssh to add an alternative to the `-c` flag that is more data-oriented. e.g. instead of:
```
setenv SSH_AUTH_SOCK /tmp/ssh-XXXXXXI8MwAa/agent.1841963;
setenv SSH_AGENT_PID 1841965;
echo Agent pid 1841965;
```
you would get:
```
SSH_AUTH_SOCK="/tmp/ssh-XXXXXXI8MwAa/agent.1841963"
SSH_AGENT_PID=1841965
```

**Comment by Notarin** [comment]
> I'm glad I helped you see the hidden strength of Nushell :-).
> 
> Aliases define declarations, similar to `def`, i.e., they add stuff to the IR. So it's not possible to define an alias coming from `zoxide init` at runtime. Doing so would also count as “eval”. I used “eval” in the above explanation for modifying parser definitions or the IR at runtime in general, it is not limited only to the `eval` keyword sourcing some code string.

@kubouch, I think you didn't quite pick up my point on alias declarations not taking effect until *after* the file evaluation and execution, loooong after IR. It should be treated as "there's alias declarations in this file and they are applied to the interactive shell session, NOT inside the same file". This would mean IR analysis will still be just as strong, if not stronger, also you already cannot tell what variables pre-exist during the initialization of a context, this would literally just not affect it at all.

Currently I think alias is applied the very line it is located in a file, which isn't how this works, and I also think it shouldn't be, but that's a big refactor haha.

So instead, how about some sort of command that it is understood that its applied aliases are effective after the current execution context expires, i.e. config finishes running or the end of someone running a command in the shell.
f.y.i. There's a lot of semantics in this, so if something I said doesn't make sense, feel free to ask me to elaborate.

**Comment by NotTheDr01ds** [comment]
@Notarin I'm not discounting your Alias ideas, but for Zoxide specifically, I think the instructions that are currently used are a bit outdated.  We now have "vendor autoload" directories where the Zoxide initialization can live.  It will be "sourced" automatically during startup.

To install Zoxide into Nushell, just run the following, one time:

```nushell
zoxide init nushell | save -f ($nu.data-dir | path join "vendor/autoload/zoxide.nu")
```

That's it.  Just restart (or source the file that was generated if you wish) and it will always get loaded during startup after that.

**Comment by Notarin** [comment]
This solution is possibly not ideal. Does the vendor auto load happen after
the config step? If so, then this works. If not, then this is still a bit
scuffed, first launch would be missing zoxide, I have all my dotfiles
synced via git, and I start new vms and installs constantly so the first
launch is important to me.

**Comment by NotTheDr01ds** [comment]
By "config step", do you mean after `config.nu` is loaded?  If so, then yes.

>  I have all my dotfiles synced via git, and I start new vms and installs constantly so the first
launch is important to me.

As long as you also add the `($nu.data-dir)/vendor/autoload` to your dotfiles repo, then yes, this is going to work on first launch.

**Comment by Notarin** [comment]
```
~ on  main [$!]
nu ❯ zoxide init nushell | save -f ($nu.data-dir | path join "vendor/autoload/zoxide.nu")
Error:   × Problem with [/home/notarin/.local/share/nushell/vendor/autoload/
  │ zoxide.nu], Permission denied
   ╭─[entry #2:1:31]
 1 │ zoxide init nushell | save -f ($nu.data-dir | path join "vendor/autoload/zoxide.nu")
   ·                               ───────────────────────────┬──────────────────────────
   ·                                                          ╰── No such file or directory (os error 2)
   ╰────
```

Seems like I'd have to implement logic ensuring the parent directory exists first I think.

**Comment by Notarin** [comment]
I want to reiterate how I believe these tools should work.
I believe that the output of these "extensions" should be a declaration of their effects, that can then be parsed and utilized by the parent nu process.
- `ssh-agent` should return a declaration of env vars and their values
- `starship init nu` should return a declaration of env vars and their values
- `zoxide init nushell` should return a declaration of aliases
Then that data is handled however the user wishes, most likely loading it into their session as env vars or aliases respectively.

I also want to reiterate why I think it's so important that the ability to load aliases from data. While nu is a pseudo-compiled lang, it is also an undeniably interactive shell, and I see the ability to procedurally manipulate shell behaviour as a vital necessity.
Another big disconnect that I think aides my point is that in an interactive shell, every line* is a new evaluation, and a whole file is one, meaning my suggested solution would provide benefit while not affecting the static nature of files.

A random thought that just popped in my head, maybe some inspiration could be taken from functional languages? Maybe making functions a first-class data type could somehow lead to a solution?

Also, is there a room, thread, or chat where this could be explored more interactively?

**Comment by jokeyrhyme** [comment]
@Notarin but there are functioning workarounds for the outcomes that want, within the "way of thinking" of nushell :)

I've got `ssh-agent`, `starship`, `atuin`, `zoxide`, `nvs`, `virtualenv`, etc all manner of pre-nushell shell integrations working just fine in nushell, we only have to let go of POSIX and mutability and `eval` and embrace what's currently available here :)

**Comment by Notarin** [comment]
Hmm, it really feels like you only read my early opinions and suggestions and either have not given my current stance a proper read or haven't quite grasped it.

**Comment by Notarin** [comment]
I wrote the following in the nushell discord, posting here as it's relevant.

# Fixing Nushell's Extensibility

## Responding to the metahack solutions document

Okay, speaking on this issue, I should really clarify my current stance (different from my first stance).
I think the issue at hand isn't this at all, and I imagine it is quite possible to have an exhaustive design model without eval capabilities.
I feel like this is possibly an XY problem.
I think if the nu team manages to grasp my intention, despite my poor linguistic skills, they will become avid supporters over eval capabilities for 95+% of relevant issues.
My idea strictly feeds into the data-oriented model of Nu(shell), while quite possibly having full compliance with the static design of Nu.

## General Solution

To state here, I believe that these types of extensions/plugins/utilities, henceforth called primarily plugins, but also interchangeably the other terms, should all operate in a strict data-oriented pipeline.
Their outputs should be what I am now coining, "A Declaration of Effects", meaning they should be parse-able to operable data (without extra non-data bits, like comments or commands. i.e. `ssh-eval -c`'s last line "`echo ...`"), and this operable data being a list of intended effects, e.g. env vars being set.

### A Show-Case of the solution

For demonstration purposes, lets have two example plugins: `simple-plugin`, and `advanced-plugin`.

#### simple-plugin

##### simple-plugin Itself

`simple-plugin` is simple, it only sets environment variables, or rather, it only returns a list of env variables with the intent of the user setting them.
When ran, `STDOUT`, including the prompt and commandline, is the following:

```stdout
nu>~ simple-plugin
foo = "whatever"
bar = "whenever"
nu>~ 
```

##### Loading simple-plugin into Nushell

This output is data oriented, parsable, and contains no incongruous comments, data, or commands. This makes it not only ideal for Nushell, and easy to utilize, but also makes it shell-agnostic, meaning everyone regardless of shell can agree on this, and not giving Nushell special treatment. It is a standard that serves everyone, not a standard forcing others to be compliant with Nushell's way.
The intended way to use `simple-plugin` is the following, most likely in `config.nu`:

```nu
simple-plugin
    | lines
    | parse "{name} = \"{value}\""
    | transpose -r
    | into record
    | load-env
```

Instead of how utilities like `ssh-agent` expect you to `eval` their output, which is shell specific, as well as inflexible, this instead is a shell-agnostic data-oriented approach. It is a declaration of effects, it just tells you *what* happens, the env vars being set, and it is flexible/parse-able enough so that it is up to whatever user in whatever shell to have, instead of a child context, their current context perform the operations in said context to operate on and utilize the data/actions.

#### advanced-plugin

##### advanced-plugin Itself

`advanced-plugin` is different. I am mentioning this to prevent a misunderstanding. This is not specifically about env manipulation, like `simple-plugin`, but rather a broader scope about "Changes", "changes" to the current shell [space] (environment is already a term for something else, context isn't quite right, and execution isn't right either, space was the closest I could think of. I hate semantics.). So, for this example plugin, it has more effects on the interactive environment, it sets env vars, aliases, and also performs a mystery function (mystery function to be explain later).
When ran, `STDOUT`, including the prompt and commandline, is the following:

```stdout
nu>~ advanced-plugin
[environment]
foo = 'who'
bar = 'what'

[alias]
buzz = 'when'
bazz = 'where'

[xxx]
qux = 'why'
quz = 'Those are the fives Ws no lie'
nu>~ 
```

This output is toml, but it could be any format as long as if the proper annotation of the purposes of each var is within the data, or at least a stable distinction is made. e.g. the following is acceptable instead:

```kv
environment.foo=who
environment.bar=what
alias.buzz=when
alias.bazz=where
xxx.qux=why
xxx.quz=Those are the fives Ws no lie
```

(This is not the format that will be used for the rest of this writing.)

##### Loading advanced-plugin into Nushell

The intended way to use `advanced-plugin` is the following, most likely in `config.nu`:

```nu
let record: record = advanced-plugin | from toml
$record.environment | load-env
$record.alias | load-into-alias-register # Does not exist nor work
for i in ($record.xxx | transpose | rename key value) {
    # Perform mystery operations
}
# OR
(advanced-plugin | from toml).environment | load-env
(advanced-plugin | from toml).alias | load-into-alias-register # Does not exist nor work
for i in ((advanced-plugin | from toml).xxx | transpose | rename key value) {
    # Perform mystery operations
}
```

Again, this is us receiving our "Declaration of Effects" and then performing them in context, live, as well as being able to change what we do with it, if we wish, with little change.

## Elaboration on What is Added

You may have noticed the presence of `load-into-alias-register`, and mystery function, I'll explain:

### load-into-alias-register

This command does not currently exist in Nu. This is a non-existent command that is used for demonstration of what a solution would look like. The simple solution of this working like `load-env` is flawed and problematic, which is why it does not yet exist. I however believe I have a solution to resolve this, and if it isn't acceptable, that there is one that exists similar in nature. I'll explain shortly why my `load-into-alias-register` does not break the pseudo-compiled nature of Nu.

### Mystery Function

I want to stress that this is **NOT** about aliases specifically, nor the `environment`. This is about changes to the interactive shell userspace, how its important to be able to manipulate it in ways not possible in scripts, and how the nature of the two differ, necessitating my proposal. This is why I wanted to include at least one more example of a type of change that would fall under the same umbrella, as well as demonstrate how something more complicated than just straight loading vars from a process into the context... but to honest with you, I couldn't think of one XD. So instead I used a pseudo function to make the example exhaustive and cover  the bases. I am sure there are more types of things that will come up in the future, and if we set this precedence then we will have a clear direction to add said issue to this solution.

### What makes load-into-alias-register Work

There are problems with just procedurally loading aliases, I mentioned this earlier, and that I would elaborate here. Nushell is a pseudo-compiled language, strengthening stability, idiosyncrasy, and static analysis, it is good. However, given this, there arises problems with aliasing procedurally, every command is a symbol, and having symbols be procedurally generated during run-time defeats pre-runtime static analysis. This is why we cannot JUST DO IT™. My solution is instead to effectively treat files as one large stage of execution, and every user-ran command (i.e. pressing enter) into a single stage execution each. Then have every `load-into-alias-register` execution bundle their results together, until the stage of execution finishes, then post-execution-stage, every entry is applied for the remaining shell session. This means at no point in a whole script file or its sources, you cannot procedurally set aliases that are effective in the same file (and sources). This means symbol resolution is untouched, intact, and the command can be fully effective to the user, allowing them to procedurally set aliases via interactive commands, their config, or even script executions. I call this bundling style of gathering changes to be applied post-execution a `register`, specifically the `alias-register`.
This is just a shower-thought, but maybe it might be worth treating more features the same? As in something like an `env-registry`?
The idea of having some separate backstage thread that applies effects *after* the command/script can sound crazy and radically different from how Nushell currently works, however, Nushell already has a very similar instance of this implemented- the `commandline edit` command. It modifies a property of the userspace, the command buffer, and this takes place *after* it executes. (I imagine this is more-so that it is actually done in the process, but it doesn't show the effects until it finishes, but I am talking about the effective result, not the technical process.)

## Alternative Solutions

I am also looking at other solutions too, one being possibly instead of having a registry, having an alias option or alternative method that only works explicitly for user typed commands, meaning it cannot *ever* modify *any* symbol in *any* scripts.
Another potential solution that is very similar, having normal aliases, but the layered on top specifically for user typed commands, sourcing a legacy layer of alias resolution from environment variables.

## What Needs to Change

I have explained my solution, but I haven't explained the difference between now and then and what needs to be done. To ensure there is not confusion about what this would look like in application, I'd like to elaborate on what conforms and what does not.

### What Nushell Needs to Change

As it stands now, Nushell is surprisingly close to being compliant with this proposal. It obviously supports executing processes, it has pipeline capabilities, it has a data-oriented approach, and it has the capability to procedurally import environment variables with `env-load`. There may be more, but currently its capabilities to procedurally edit userspace is limited to solely just the environment variables, unless there is more I know not of.
While Nushell is surprisingly close to compliance, it is still potentially far off. There is an unknown number of facets to the userspace, I can immediately name two, the environment variables, and aliases, of which Nushell only complies with one, env vars. This means Nushell lacks compliance with one, or possibly many more, facet of my proposed solution. Nushell cannot procedurally create aliases. This is absolutely required to create a, IMO, sane data-oriented model for extensive functionality.

### What ssh-agent Needs to Change

Right now, `ssh-agent` is *not* compliant with my proposed standard. It has three different output formats, and not a single one is compliant. The options being no flags, `-c`, and `-s`, which are all unacceptable. Every single one of these are firstly, not shell-agnostic, and secondly, not data-oriented. The shell favoritism is problematic for pretty obvious reasons, it favors specific shells and does not provide a single non-shell-specific output, meaning if you use any shell that is not lucky enough to happen to be able to parse the output, then you will have to reverse the shell specific output back into shell-agnostic data, and then from there build it into something usable by your shell. I see no issue with opt-in shell specific compatibility layers, but having absolutely no option for any of the other thousands+ is an issue to me, all it takes is one *default* option. The suggested solution demonstrates the parse-gymnastics as shown below, thankfully due to Nushell being super flexible it is little work:

```nu
^ssh-agent -c
    | lines
    | first 2
    | parse "setenv {name} {value};"
    | transpose -r
    | into record
    | load-env
```

The lack of data-orientation in the output is also on its own an issue. `ssh-agent`'s output elects an executive-oriented model instead, outputting explicit, inflexible, and unstable commands, to be evaluated by a hopefully compliant shell. Heaven forbid what happens if you just so happened to, for whatever reason, alias a command that is in the output, which *cough* wouldn't be a problem in a data oriented model. I've also yet to mention the absolute *absurdity* of it having it having an `echo` command baked in! Why?! In the effective space of the command execution, its purpose is *just* to apply technical data to the env, it does so, and then it stupidly prints out the last bit of data instead of also storing it in the env and allowing the user to handle it as they please. This means if the user wishes to change how that is handled, e.g. store that as a variable, they will have to reverse engineer the result, either reverse-parsing the command result, or the eval result. The offensive output is below:

```nu
nu>~ ssh-agent -c
setenv SSH_AUTH_SOCK /tmp/ssh-XXXXXXvKxLD0/agent.3434117;
setenv SSH_AGENT_PID 3434119;
echo Agent pid 3434119;
nu>~ 
```

Fixing both core issues, and achieving compliance would look like the following:

```nu
nu>~ ssh-agent -c
SSH_AUTH_SOCK /tmp/ssh-XXXXXXvKxLD0/agent.3434117;
SSH_AGENT_PID 3434119;
nu>~ 
```

It is very similar, having to change very little to achieve compliance.

### What Zoxide Needs to Change

Zoxide is a **very** non-compliant, its output is so executive-oriented and messy I won't even bother showing it. But all it does is set environment variables, custom commands, and aliases. The custom commands can likely be coerced into aliases, meaning all it needs to achieve compliance is just output everything as declarative data, and thus be shell-agnostic and puts more power in the users hands to handle the data how they please.

### What Starship Needs to Change

Starship is *incredibly* close to compliance, literally *all* it outputs is environment variables, meaning it can, with very little effort, just output the exact same thing, just as KV data.

## The Idea This is a Non-Issue

I've also entertained the idea that this could be argued to be a non-issue, and that leaning into the pseudo-compiled nature of Nushell would look more like instead rewriting those tools in Nu. I think that this is an understandable idea, but it is naive and dismissive of the difficulty of actually doing so.

## How The Official Docs are an Anti-Pattern

I believe that the current recommended method for handling plugins is an anti-pattern. The cited issue on why procedural aliases are not easy to add is because static analysis needs to be protected, however, the publicly recommended alternative is actually *just* as bad!
The way it is now, if you design a script to write a nu file, then source it, the file will not exist for the language server to analyze, meaning it completely fails to operate as expected. Even if the file exists already, it *will* be overwritten next launch of the script. The current analysis of the file will *always* expire before the code is ran, meaning you will *never* have a guaranteed correct analysis, and it will only ever be correct out of chance, chance that the next generation will be identical to the previous one.

## Final Note

I wanna just follow this up with an explanation that the solutions and ideas I have presented are not what I expect to be implemented. There is obviously, to some degree, no matter how small, some flaw to it. I want this to be a jumping point that indicated what I believe is the ideal direction to take in achieving the solution, and for the details to be worked out communicatively.

**Comment by NotTheDr01ds** [comment]
> Seems like I'd have to implement logic ensuring the parent directory exists first I think.

Ah, sorry - I do often forget that part.  The logic is very straightforward.  Just change my instructions above to:

```nushell
($nu.data-dir | path join "vendor/autoload/") | mkdir $in
zoxide init nushell | save -f ($nu.data-dir | path join "vendor/autoload/zoxide.nu")
```

Nushell won't error if the directory already exists, so it's fine to execute in either state.

**Comment by Notarin** [comment]
I made some assertions I would work on stuff and update some stuff, but unfortunately I fell ill, as well as busy with the otherwise.
So, oops, and sorry.

To make a quick update so I don't leave things totally hanging:
I have changed my mind on some things.
I have spoke with some contributors in the discord, and while I stand on much of the spirit of the original proposal, I now hold a very boldly opposite stance on some points, for differing reasons.
I am **strongly** against being able to procedurally declare aliases or commands.
I believe that everything to do with custom-commands and aliases should be declaratively defined by the user, **not** third-party developers.
I am so boldly against procedural aliases that I not hold the very *opposite* stance on the current official solutions/workarounds.
I hate them for the fact they are technically procedural in nature. They run commands, which write to files, and then source the files. I hate this because this means the aliases and commands are defined by the third party developer, and can be changed or be anything at any time, without user intervention.
I despise this for several reasons:
1. POLS. This is non-compliant with the POLS guideline, Principle of Least Surprise. It is too possible that an alias or command can conflict with something pre-existing in the users workflow, disrupting it, and "surprising" the user. This can lead to arcane and difficult to debug situations, as well as irreversible unintended destructive actions for users.
2. Stability. This is a bit or a more difficult to explain point, but procedural aliases is unstable. If you find yourself needing it, you are the victim of an XY problem, your approach is poor, and you should reevaluate it in search of the solution, Z. Also, it is functionality of the past, like eval, this functionality belongs to less stable and complex nuanced shells of yesteryear ( see you in hell bash delimiters >:( ).
3. Security. It is possible for destructive and unintentional aliases or commands to be set, possible through malicious means, but even more possible and importantly, possibly through non-malicious means.
Let's say Timmy has an alias, `zi`, and Timmy installs zoxide (pretend the zoxide `zi` doesn't exist), everything's well until at one point in the future zoxide broadens its functionality, it adds new commands. It adds an install alias for every shell it supports, `bi`, `ei`, `fi`, `ni`, `pi`, `xi`, and `zi`. It adds these aliases to the procedurally generated alias file, read by the shell, in this case Nushell.
See theres no malice, it was designed to be helpful! In the developers eyes they just wanted to make things easier for users! Then when the repo maintainers reviewed the changes, they thought the same, "this is a good change and will make things easier for the user!". But none of them know little old Timmy has an alias called 'zi'. Even worse, this feature's new, each alias writes dozens on dozens of files and configuration changes, and the developer forgot to make an uninstall command!
Well the change ships, and little poor Timmy runs 'zi', and then his zsh config is broken, his home partition is polluted, as well as dozens on system config changes were made. Timmy has no clue what just happened, "weird, it shouldn't have done that, it ran, spat nothing out to `STDOUT`, and returned with a 0 exit status...", so Timmy runs it several more times, as Timmy did what most people would do, and multiplies the damage. Later Timmy discovers their ZSH config is nuked, and it's due to dozens of files and config changes in random places. Timmy has to spend a lot of time digging to find out what happened, turns out zoxide just recently released a `zi` alias! Well that explains everything! Timmy makes an issue asking how to undo the command, and the dev responds with bad news, they forgot to add an uninstall alias and they're on vacation for the next two weeks... So here's Timmy manually sifting through hundreds, if not thousands of files, removing the new files, sifting through their ZSH config, undoing `zi`'s changes, and sifting through the system config, undoing all options changed.
This isn't even the most extreme possibility, through some crazy leaps, and really odd luck, things like an accidental `rm -rf --no-preserve-root /` could happen.
While these are extreme examples, the fact they are even possible is an issue.

It's rather ironic and funny, because I initially had a very opposite stance, and the people I was talking with, I differed in opinion with, upon speaking with them and further thought into it, I now hold the radically opposite stance, and disagree with them *again*, just differently! I will die on this hill. XD

To all those who previously supported me, sorry, but I have turned to the dark side XP

**Comment by ikcikoR** [comment]
What about:
```bash
exec bash -c "your-bash-only-thing && exec nu"
```

**Comment by rsenna** [comment]
I'm a seasoned developer, but a new Nushell user.
And I wasn’t fully aware Nushell had this “issue”… which, according to the “true believers,” isn’t an issue at all but a deliberate design choice.

I only discovered it today while finishing the migration from my fish-based setup to Nushell.  
Things broke immediately, because *many* developer tools I use require dynamic script generation + `eval` to work.

Yes, I noticed there are workarounds that “respect” Nushell’s statically-compiled purity, and I understand the team’s goal of encouraging a static parsing mindset. It’s a noble aim.

But in practice, it means Nushell will be used *less often* and *less elegantly* than it could be.

For developers like myself (and many other IT professionals) the *primary* purpose of a shell is to *run dynamic commands*. A shell interpreter is a launchpad for other CLI tools; it’s **not** the end goal!

So, despite the good intentions, many users will keep resorting to ugly hacks, just like the comment before mine: `exec bash`, run the “impure” stuff there, then `exec nu` again. That's sure *ugly*, even if it works!

As a developer, I believe that when restrictions push me toward *worse* code rather than more elegant solutions, it’s time to *reconsider* those restrictions. Rust is famously strict, but it still allows developers to bypass memory safety through `unsafe` blocks. So, please, **be like Rust**. Why not offer an *eval mode* toggle, perhaps as a shebang-line argument, to explicitly mark a script as *unsafe* for static guarantees…? That would be a reasonable compromise, giving people the flexibility to use Nushell in the way that best fits their needs.

**Comment by fdncred** [comment]
I've seen an eval patch to add a command called 'evil' that does some of this (saw it on [discord](https://discord.com/channels/601130461678272522/601130461678272524/1390704640529530900)). I also have a [nu_plugin_run_internal](https://github.com/fdncred/nu_plugin_run_internal) that does some of it. I'm not against an eval command but it's definitely against the philosophy of nushell.

**Comment by Notarin** [comment]
@rsenna 
I really disagree with a lot of this.
nushell is turing complete, it doesnt disallow any computation, you just may have to take a more declarative approach, which is more secure, type safe, and sane.
You can absolutely still have arbitrary behaviour, but you're going to have to unlearn some things that are simply impossible to do in nushell.
A huge purpose of nushell is to have reasonable expectations about what a value is and how it behaves. Allowing for even alittle undefined behavior would warrant a massive rewrite, and in the end you would just end up with another bash with different rules.

"many users will keep resorting to ugly hacks" A developer cannot stop a user from doing dumb things, nor should they care.

"when restrictions push me toward worse code rather than more elegant solutions" maybe instead of reconsidering the philosophy, reconsider the tool? It might just not be for you, not all good things are meant to be for everyone, sometimes a good thing just isnt something that works for you.

"Why not offer an eval mode toggle, perhaps as a shebang-line argument, to explicitly mark a script as unsafe for static guarantees…?" Because this isn't the case and it isnt a job for the tool.
Adding such a mode would be adding a drill bit to a hammer... why would you? Just use a screwdriver.

"giving people the flexibility to use Nushell in the way that best fits their needs" Again, this only stretches so far.

Yes, some people are using ugly hacks, but that doesn't mean a philosophically pure method of doing what they want doesn't exist. That would be like blaming python developers for users misusing the language to abuse features.

Commands are citizens for data as well, and giving them this arbitrary space where anything could be a command would nuke the entire purpose of the nu interpreter.

I want to stress that just because you don't know a "Nu" way to do something, doesn't guarantee a reasonable way doesn't exist.

**Comment by jokeyrhyme** [comment]
The shell output from various tools targets a POSIX-compatible `sh`

And nushell so far is intentionally not POSIX compatible, just like fish and elvish and many other shells

In order to be out-of-the-box compatible with every existing tool, nushell would effectively have to become POSIX-compatible, which would not only have implications for the mutability model, but would also mean including a Rust implementation of `sh`

However, what I've noticed lately is that many newer tools (and even some older ones) are offering first-party support for non-POSIX shells

So, perhaps the proliferation of non-POSIX shells is a forcing function on the terminal/shell ecosystem, and perhaps tools that don't become sufficiently shell-agnostic will eventually see less and less usage over time?

**Comment by Notarin** [comment]
> What about:
> 
> exec bash -c "your-bash-only-thing && exec nu"

This is what shebangs are for. Just add a shebang to your script and then run `./script.sh`


EDIT: Even better, add the shebang, remove the extension, throw it in `.local/bin/`, and tadaaaa, you now have it as a command like every other application you have insalled.
